SecurityDefinition oauth2 fails with type mismatch

When using securitDefition for oauth2 like specified in the readme:
https://github.com/gigaSproule/swagger-gradle-plugin/blob/master/README.md#oauth2---security-definitions

it fails with 

>  Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize instance of `java.util.LinkedHashMap` out of START_ARRAY token
     at [Source: UNKNOWN; line: -1, column: -1] (through reference chain: io.swagger.models.auth.OAuth2Definition["scopes"])
...
at com.benjaminsproule.swagger.gradleplugin.swagger.SwaggerFactory.getSecuritySchemeDefinitionByType(SwaggerFactory.groovy:173)

There seems a mismatch between 
 `com.benjaminsproule.swagger.gradleplugin.model.SecurityDefinitionExtension` and `io.swagger.models.auth.OAuth2Definition` it tries to create (intermediate is the json creation).

Will analyse it later.

--- a/src/main/groovy/com/benjaminsproule/swagger/gradleplugin/model/SecurityDefinitionExtension.groovy
+++ b/src/main/groovy/com/benjaminsproule/swagger/gradleplugin/model/SecurityDefinitionExtension.groovy
@@ -34,4 +34,12 @@ class SecurityDefinitionExtension {
     void scope(Closure closure) {
         scopes += project.configure(new ScopeExtension(), closure) as ScopeExtension
     }
+    
+    List<ScopeExtension> getScopeExtensions() {
+      return scopes;
+    }
+    
+    Map<String,String> getScopes() {
+      return scopes.collectEntries { [ (it.name) : it.description ] }
+    }
 }

--- a/src/main/groovy/com/benjaminsproule/swagger/gradleplugin/validator/SecurityDefinitionValidator.groovy
+++ b/src/main/groovy/com/benjaminsproule/swagger/gradleplugin/validator/SecurityDefinitionValidator.groovy
@@ -65,10 +65,10 @@ class SecurityDefinitionValidator implements ModelValidator<SecurityDefinitionEx
             errors += 'securityDefinition.authorizationUrl is required by the swagger spec for OAuth 2.0'
         }
 
-        if (!securityDefinitionExtension.scopes) {
+        if (!securityDefinitionExtension.getScopeExtensions()) {
             errors += 'securityDefinition.scopes is required by the swagger spec for OAuth 2.0'
         } else {
-            securityDefinitionExtension.scopes.each { scope ->
+            securityDefinitionExtension.getScopeExtensions().each { scope ->
                 errors += scopeValidator.isValid(scope)
             }
         }

--- a/src/test/groovy/com/benjaminsproule/swagger/gradleplugin/SecurityITest.groovy
+++ b/src/test/groovy/com/benjaminsproule/swagger/gradleplugin/SecurityITest.groovy
@@ -7,21 +7,21 @@ import static org.gradle.testkit.runner.TaskOutcome.SUCCESS
 import java.util.List
 
 class SecurityITest extends AbstractPluginITest {
-  
-    static final List<String> springMvcAndJaxRs = [
-       """
+
+  static final List<String> springMvcAndJaxRs = [
+    """
           springmvc = true
           locations = ['com.benjaminsproule.swagger.gradleplugin.test.springmvc.SampleController']
        """,
-       """
+    """
           locations = ['com.benjaminsproule.swagger.gradleplugin.test.jaxrs.SampleResource']
        """
-      ]
-    
-    def 'Configure global security'() {
-      given:
-      def expectedSwaggerDirectory = "${testProjectOutputDirAsString}/swaggerui-" + UUID.randomUUID()
-      buildFile << """
+  ]
+
+  def 'Configure global security'() {
+    given:
+    def expectedSwaggerDirectory = "${testProjectOutputDirAsString}/swaggerui-" + UUID.randomUUID()
+    buildFile << """
             plugins {
                 id 'java'
                 id 'groovy'
@@ -46,24 +46,24 @@ class SecurityITest extends AbstractPluginITest {
             }
         """
 
-      when:
-      def result = runPluginTask()
+    when:
+    def result = runPluginTask()
 
-      then:
-      result.task(":${GenerateSwaggerDocsTask.TASK_NAME}").outcome == SUCCESS
+    then:
+    result.task(":${GenerateSwaggerDocsTask.TASK_NAME}").outcome == SUCCESS
 
-      assertSwaggerJsonGlobalSecurity("${expectedSwaggerDirectory}/swagger.json")
+    assertSwaggerJsonGlobalSecurity("${expectedSwaggerDirectory}/swagger.json")
 
-      where:
-      testSpecificConfig << springMvcAndJaxRs
+    where:
+    testSpecificConfig << springMvcAndJaxRs
   }
-  
+
   private static void assertSwaggerJsonGlobalSecurity(String swaggerJsonFile) {
     def producedSwaggerDocument = new JsonSlurper().parse(new File(swaggerJsonFile), 'UTF-8')
 
     assert producedSwaggerDocument.swagger == '2.0'
     assert producedSwaggerDocument.basePath == '/'
-    
+
 
     def info = producedSwaggerDocument.info
     assert info
@@ -74,12 +74,12 @@ class SecurityITest extends AbstractPluginITest {
     assert security
     assert security.size() == 1
     assert security[0].containsKey('MyBasicAuth')
- }
- 
- def 'Configure global security with multiple auths'() {
-   given:
-   def expectedSwaggerDirectory = "${testProjectOutputDirAsString}/swaggerui-" + UUID.randomUUID()
-   buildFile << """
+  }
+
+  def 'Configure global security with multiple auths'() {
+    given:
+    def expectedSwaggerDirectory = "${testProjectOutputDirAsString}/swaggerui-" + UUID.randomUUID()
+    buildFile << """
             plugins {
                 id 'java'
                 id 'groovy'
@@ -103,8 +103,7 @@ class SecurityITest extends AbstractPluginITest {
                         type = 'apiKey'
                         keyLocation = 'header'
                         keyName = 'X-API-Key'
-                    }
-                    /* Add example when  issue https://github.com/gigaSproule/swagger-gradle-plugin/issues/104 is fixed 
+                    } 
                     securityDefinition {
                         name = 'MyOAuth2'
                         type = 'oauth2'
@@ -115,41 +114,66 @@ class SecurityITest extends AbstractPluginITest {
                             description = 'description'
                         }
                     }
-                    */
-                    security = [ [ MyBasicAuth : [], MyApiKey : [] ] ]
+                    security = [ [ MyBasicAuth : [], MyApiKey : [] ], [ MyOAuth2 : [ 'scope1' ] ] ]
                     ${testSpecificConfig}
                 }
             }
         """
 
-   when:
-   def result = runPluginTask()
+    when:
+    def result = runPluginTask()
 
-   then:
-   result.task(":${GenerateSwaggerDocsTask.TASK_NAME}").outcome == SUCCESS
+    then:
+    result.task(":${GenerateSwaggerDocsTask.TASK_NAME}").outcome == SUCCESS
 
-   assertSwaggerJsonGlobalSecurityMultiple("${expectedSwaggerDirectory}/swagger.json")
+    assertSwaggerJsonGlobalSecurityMultiple("${expectedSwaggerDirectory}/swagger.json")
 
-   where:
-   testSpecificConfig << springMvcAndJaxRs
- }
- 
- private static void assertSwaggerJsonGlobalSecurityMultiple(String swaggerJsonFile) {
-   def producedSwaggerDocument = new JsonSlurper().parse(new File(swaggerJsonFile), 'UTF-8')
+    where:
+    testSpecificConfig << springMvcAndJaxRs
+  }
 
-   assert producedSwaggerDocument.swagger == '2.0'
-   assert producedSwaggerDocument.basePath == '/'
+  private static void assertSwaggerJsonGlobalSecurityMultiple(String swaggerJsonFile) {
+    def producedSwaggerDocument = new JsonSlurper().parse(new File(swaggerJsonFile), 'UTF-8')
 
-   def info = producedSwaggerDocument.info
-   assert info
-   assert info.version == '1'
-   assert info.title == 'test'
+    assert producedSwaggerDocument.swagger == '2.0'
+    assert producedSwaggerDocument.basePath == '/'
 
-   def security = producedSwaggerDocument.security;
-   assert security
-   assert security.size() == 1
-   assert security[0].containsKey('MyBasicAuth')
-   assert security[0].containsKey('MyApiKey')
- }
- 
+    def info = producedSwaggerDocument.info
+    assert info
+    assert info.version == '1'
+    assert info.title == 'test'
+    
+    // Validate securityDefinitions
+    def securityDefinitions = producedSwaggerDocument.securityDefinitions;
+    assert securityDefinitions
+    assert securityDefinitions.size() == 3
+    
+    def basicAuthDef = securityDefinitions['MyBasicAuth']
+    assert basicAuthDef
+    assert basicAuthDef.type == 'basic'
+    
+    def apiKeyDef = securityDefinitions['MyApiKey']
+    assert apiKeyDef
+    assert apiKeyDef.type == 'apiKey'
+    
+    
+    def oauth2Definition = securityDefinitions['MyOAuth2']
+    assert oauth2Definition
+    assert oauth2Definition.type == 'oauth2'
+    assert oauth2Definition.authorizationUrl == 'authorizationUrl'
+    assert oauth2Definition.flow == 'implicit'
+    assert oauth2Definition.scopes
+    assert oauth2Definition.scopes['scope1'] == 'description'
+    
+    
+    // Validate security
+    def security = producedSwaggerDocument.security;
+    assert security
+    assert security.size() == 2
+    assert security[0].containsKey('MyBasicAuth')
+    assert security[0].containsKey('MyApiKey')
+    assert security[1].containsKey('MyOAuth2')
+    
+  }
+
 }

